server/server.go

```diff
@@ -82,14 +80,14 @@ func NewShipperServer(cfg Config, publisher Publisher) (ShipperServer, error) {
 	return &s, nil
 }
 
-// GetAcceptedIndex returns the accepted index
-func (serv *shipperServer) GetAcceptedIndex() uint64 {
-	return uint64(serv.publisher.AcceptedIndex())
-}
-
 // GetPersistedIndex returns the persisted index
 func (serv *shipperServer) GetPersistedIndex() uint64 {
-	return uint64(serv.publisher.PersistedIndex())
+	index, err := serv.publisher.PersistedIndex()
+	if err != nil {
+		// An error means the queue has been closed, so just return a placeholder.
+		return 0
```

Comment:
Can we just change the return type to `(uint64, error)` here instead? Zero is a valid index so it feels wrong to return it when the return value should really be invalid.

Reply:
I tried, and it's possible but inconvenient because some callers don't really have anyone to report an error to. But this is really more of a local helper than API, so instead I made it private and documented that you should use a direct queue call if you want to explicitly recognize queue closure, is that a good alternative?

Reply:
That's fine, my main concern is an invalid index of 0 inadvertently being observed a client as the shipper is shutting down such, if not now then maybe eventually as this code changes.

If you think that's low risk or impossible I'm fine with this as is.

Reply:
Seeing code like this below makes me worry this might actually be possible if the queue can shut down/close before the gRPC server does. I'm not sure we enforce an order of operations anywhere.

```go
	persistedIndex := serv.getPersistedIndex()
	err := producer.Send(&messages.PersistedIndexReply{
		Uuid:           serv.uuid,
		PersistedIndex: persistedIndex,
```

I would prefer something like this as a guard against this:

```go
	persistedIndex, err := serv.getPersistedIndex()
    if err != nil {
        return status.Error(codes.Unavailable, err)
    }
	err := producer.Send(&messages.PersistedIndexReply{
		Uuid:           serv.uuid,
		PersistedIndex: persistedIndex,
```

If the shipper is about to shutdown the RPC calls should either complete successfully in an orderly manner or return an explicit error to abort them. I suppose we could also just let the stream be torn down without returning anything here.

Reply:
I think this should be sufficient:

```diff
diff --git a/server/server.go b/server/server.go
index 716ce97..1406fc0 100644
--- a/server/server.go
+++ b/server/server.go
@@ -195,7 +195,7 @@ func (serv *shipperServer) PersistedIndex(req *messages.PersistedIndexRequest, p
 
 		case <-ticker.C:
 			newPersistedIndex := serv.GetPersistedIndex()
-			if newPersistedIndex == persistedIndex {
+			if newPersistedIndex <= persistedIndex {
 				continue
 			}
 			persistedIndex = newPersistedIndex
```

in https://github.com/elastic/elastic-agent-shipper/blob/34571c41e892f8498d817a3b2784cc372b0985a2/server/server.go#L197-L205

Reply:
This would break when we wrap around from the maximum uint64 value back to zero wouldn't it? This would only matter if the indexes are persisted in the disk queue though, I don't think the shipper is will run long enough to see this happen if the indexes are only in memory.

Reply:
Well, we could ingest 4 billion events a second for 40 years before wrapping became a problem in practice... but I do prefer not to rely on coincidences of type to resolve state transitions, so I changed this instance to call `serv.publisher.PersistedIndex` directly and check for errors, and removed one other optional call. Now `getPersistedIndex` is a local helper with a warning, _and_ all its remaining callers actually want to fall back on 0 when there's no queue.